### ISSUE TITLE
Dependency conformance tests - test libs vs solidity implementations

### ISSUE THREAD

**Issue by pdobacz** [omisego/elixir-omg#901]
Taken from "OMG-394 - Exhaustive (possibly propchecky) tests of essential crypto and libs"

Some essential crypto-code we use in prod is tested quite thinly.

Let's make an exhaustive test suite which:
1/ generates some input data to the respective libs
2/ does Elixir calculations on it (invoked via some of our public APIs like `Transaction.hash`, `Transaction.encode`, `Block.hashed_txs_at` but effectively boiling down to dependency calls
3/ validates these calculations with Elixir code
4/ redos/validates these calculations against some public API of a `solc`-library (like `Merkle.sol` or `RLP.sol`) we use - deployed lightly to a `geth --dev` instance

test data generation can be manual, but this is perhaps a good (and much easier than state machine testing we did for `State.Core`) application of PropCheck

The candidates for such tests:
A/ `Block.hashed_txs_at` (invoking `merkle_tree_elixir`) vs `Merkle.sol` - correct proofs work and incorrect don't, for "all" imaginable combinations
B/ encoding/decoding https://github.com/omisego/security-issues/issues/2
C/ signatures

**Motivation** and parent issue is #853

EDIT - crossed RLP out
EDIT2 - uncrossed RLP back in (changed to generic encoding/decoding) :joy:

**Comment by InoMurko** [comment]
If I understand correctly, you want to test if (for example) our dependency https://github.com/mana-ethereum/ex_rlp works as intended against `RLP.sol`?

**Comment by InoMurko** [comment]
Would this go under https://github.com/omisego/elixir-omg/issues/853 @Pongch ?

**Comment by pdobacz** [comment]
ah I forgot to mention #853, yes, adding to the desciption

**Comment by InoMurko** [comment]
EX_RLP is running tests against Ethereum common tests https://github.com/ethereum/tests/tree/04940788157c88c75b014e5e6969a249f49db4c0 
not sure what coverage are we going to gain by adding tests against `RLP.sol`?

**Comment by pdobacz** [comment]
> EX_RLP is running tests against Ethereum common tests

ah I wasn't aware of that (but kinda could've expected that). Thanks @InoMurko! that's one off the list or at least 1/2 off the list as:

That would leave `RLP.sol` to check - do we know whether it has the same coverage? @pik694 @boolafish I think you've mentioned touching RLP libs recently. What is the implementation of RLP that's going to be used in the long term for `plasma-contracts`?

**Comment by boolafish** [comment]
https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/contracts/src/utils/RLP.sol

solidity 0.5 version. we've added some basic test on that lib, but at the same time, only the really minimum basic at this moment.

**Comment by pdobacz** [comment]
Thx @boolafish 

Ok still clone&own. So how about this: let's expand the `plasma-contracts` test suite with tests that use the https://github.com/ethereum/tests/tree/04940788157c88c75b014e5e6969a249f49db4c0 (or more current version?) similar to what `ex_rlp` does. I'll add a separate issue there.

Then, to me, this will be good enough to take RLP off the list - the e2e tests will be enough.

**Comment by boolafish** [comment]
sure, if there is good test case we can copy it would be nice. There is just no good library (or just I haven't found one) to use for RLP that's why we're still clone&own (and....bumping version, awwwh) on this.

**Comment by pdobacz** [comment]
No worries, I was just figuring out where to file the issue. When we feel like it, we can roll RLP.sol out as a standalone lib, sometime (it's a clone, so some proper open-source-friendly process would apply ofc)

**Comment by InoMurko** [comment]
a comment for 
```
A/ Block.hashed_txs_at (invoking merkle_tree_elixir) vs Merkle.sol - correct proofs work and incorrect don't, for "all" imaginable combinations
```
If we are unsure of the dependency "correctness" we should fork it and introduce extended tests there and push it upstream. It seems to me it makes the code cleaner. The alternative is that our main repo does the testing of the dependency correctness. Which might become bloated over time. If you guys agree of the approach I'd tackle this. cc @achiurizo

**Comment by pdobacz** [comment]
> If we are unsure of the dependency "correctness" we should fork it and introduce extended tests there and push it upstream

hmm, I'd say not only "correctness" matters but also "sameness" (or conformance, as title says). A lot depends on the parameters used when invoking `merkle_tree_elixir`. More coverage on the dependency itself (and not only `merkle_tree_elixir` but `Merkle.sol` too!) wouldn't hurt for sure, but I'd sleep better if we did test a slightly higher level function `Block.hashed_txs_at` and compare with specifically the solidity implementation.

For now the "sameness" is covered only for very simple cases of probably 1-leaf trees, covered by integration tests that run against the contract.

This is different than RLP, as there there was a much cleaner/less ambiguous API and a sturdy reference test suite.

Also pragmatically - such conformance test would be immune to us changing the library used (on either end) and seems just easier to deliver than expanding two suites of tests (easier to leverage e.g. `propcheck`, since our property is simple - proof generated in one must verify in the other).

**Comment by achiurizo** [comment]
> Some essential crypto-code we use in prod is tested quite thinly.

 When you say `crypto-code`, are you referring to the third party dependencies or the `code` that wraps/leverage their behavior?

> 1/ generates some input data to the respective libs
> 2/ does Elixir calculations on it (invoked via some of our public APIs like Transaction.hash, > Transaction.encode, Block.hashed_txs_at but effectively boiling down to dependency calls
> 3/ validates these calculations with Elixir code

For me, things like `Block.hashed_txs_at` having unit tests would be **awesome** and I think these points can be handled by having those.  Having these kinds of unit tests also makes it great for us to check our expectations of 3rd party library and what our code does with it. Also makes refactoring a lot more manageable.

> 4/ redos/validates these calculations against some public API of a solc-library (like Merkle.sol or RLP.sol) we use - deployed lightly to a geth --dev instance

For third party libraries specifically, if we are testing its behavior and ensuring its 'correctness', it doesn't belong in the `elixir-omg` and I would opt towards @InoMurko's suggestion of forking and pushing changes upstream if they so desire them.

**Comment by pdobacz** [comment]
Hm, let me maybe clarify my point a bit more:

> things like Block.hashed_txs_at having unit tests would be awesome

But what would such unit tests look like? You either test against test vectors (not enough, brittle, there's no standard that I know of) or you test with proof validation from within the same lib (a bug/discrepancy might be there, but is missed because you're testing against the tested code).

We do as much unit testing as I think is reasonable:
 - single smoke test against (our own) test vector, just to pick up evident change of result
 - test correct format & response size, just in case
 - (and of course, unit tests inside the respective libs, which satisfy exactly the "proof validation from within the same lib" part)

I think you cannot go much further than this, because the inputs/outputs are basically random-looking data (merkle roots and inclusion proofs).

The behavior we're actually after testing here is the **sameness** (not only correctness, again). We want the merkle proofs produced in elixir-omg to verify correctly in solidity. (And this behavior is completely independent from our choice of the dependencies/libs we actually use.)

And to the last point from your comment (services working together), our integration tests are testing that, but very "heavily" and only in very simple cases (1-leaf-tree). We have absolutely no guarantees that the merkle proofs will validate in the contract for, say, 2^16 leaf-tree (full).

**Comment by InoMurko** [comment]
If that's the case we should adopt `Ethereum style testing` (which is common everywhere else as well).
We should specify tests that are indifferent to the implementation (inputs/outputs) in a text file ~ json ~. We put that in a repository that all contributors in our system (elixir-omg, plasma contracts, etc.) need to `pull in` and run tests against them.

**Comment by pdobacz** [comment]
> We should specify tests that are indifferent to the implementation (inputs/outputs) in a text file ~ json ~. We put that in a repository that all contributors in our system (elixir-omg, plasma contracts, etc.) need to `pull in` and run tests against them.

Aye, that would be an alternative.

IMHO, I still think that's more work (1 largish json file to design and produce data for + 2 test runners that read that json file) and more maintenance than doing the property test in `elixir-omg` (or a more rudimentary auto-generated test case) and just asserting all generated proofs validate. But not 100% sure on that.

**Comment by pdobacz** [comment]
Hm, I think we missed putting notes after one call ~month ago

I'm fairly sure, that at least in the domain of 

> A/ Block.hashed_txs_at (invoking merkle_tree_elixir) vs Merkle.sol - correct proofs work and incorrect don't, for "all" imaginable combinations

we did end up deciding that the originally proposed "conformance test approach; `ex` vs `sol` in an auto-generated test suite" is viable? Does anyone else here have the same recollection?

**Comment by paulperegud** [comment]
> EX_RLP is running tests against Ethereum common tests

Unfortunately it is still possible to decode two binaries into identical term. See here: https://github.com/mana-ethereum/ex_rlp/pull/24

**Comment by paulperegud** [comment]
And a followup - a fuzzer for Solidity library. Also, just a test, no fix provided (yet?). It fails in the same way as ex_rlp.

https://github.com/omisego/elixir-omg/pull/1135

**Comment by pdobacz** [comment]
just making a note that encoding/decoding conformance testing is back on the table (since the official RLP tests didn't cover for some issues uncovered).

Will edit the description, tracking this effort in https://github.com/omisego/security-issues/issues/2

**Comment by pdobacz** [comment]
Encoding and EIP712 conformance solved in #1164. Merkling remains from the list. Do you think I take this over @InoMurko, as a quick follow-up to #1164?
